include/libhal-soft/i2c_bit_bang.hpp

```diff
@@ -0,0 +1,210 @@
+   * @param p_timeout a timeout function which is primarily used for clock
+   * stretching to ensure the peripheral doesn't hold the line too long
+   */
+  void write(std::span<const hal::byte> p_data_out,
+             function_ref<hal::timeout_function> p_timeout);
+
+  /**
+   * @brief this function will handle writing a singular byte each call while
+   * also retrieving the acknowledge bits
+   *
+   * @param p_byte_to_write this is the byte that will be written to the bus
+   * @param p_timeout a timeout function which is primarily used for clock
+   * stretching to ensure the peripheral doesn't hold the line too long
+   * @return bool - true when the byte written was ack'd and false when it was
+   * nack'd
+   */
+  bool write_byte(hal::byte p_byte_to_write,
+                  function_ref<hal::timeout_function> p_timeout);
+
+  /**
+   * @brief this function will write a single bit at a time, dealing with
+   * simulating the clock and the clock stretching feature
+   *
+   * @param p_bit_to_write the bit which will be written on the bus
+   * @param p_timeout a timeout function which is primarily used for clock
+   * stretching to ensure the peripheral doesn't hold the line too long
+   */
+  void write_bit(hal::byte p_bit_to_write,
+                 function_ref<hal::timeout_function> p_timeout);
+
+  /**
+   * @brief this function will read in as many bytes as allocated inside of the
+   * span while also acking or nacking the data
+   *
+   * @param p_data_in a span which will be filled with the bytes that will be
+   * read from the bus
+   * @param p_timeout a timeout function which is primarily used for clock
+   * stretching to ensure the peripheral doesn't hold the line too long
+   */
+  void read(std::span<hal::byte> p_data_in,
+            function_ref<hal::timeout_function> p_timeout);
+
+  /**
+   * @brief this function is responsible for reading a byte at a time off the
+   * bus and also creating the byte from bits
+   *
+   * @return hal::byte the byte that has been read off of the bus
+   */
+  hal::byte read_byte();
+
+  /**
+   * @brief this function is responsible for reading a single bit at a time off
+   * of the bus while also managing the clock line. It will release (pull sda
+   * high) every time it is called
+   *
+   * @return hal::byte
+   */
+  hal::byte read_bit();
+
+  /**
+   * @brief this function is a high speed version of the hal::delay function
+   * which operates on ticks rather then a time duration where hal::delay
+   * provides delays up to 17us, this function will provide delays up to
+   * about 2.7us making it optimal for high accuracy delay
+   *
+   * @param ticks the amount of ticks this function will delay for, this should
+   * be based on the frequency of the clock passed into the bit_bang class
+   */
+  [[gnu::always_inline]] inline void high_speed_delay(uint64_t ticks);
+
+  /**
+   * @brief an output pin which will mimic the behavior of an I2C scl pin
+   */
```

Comment:
```suggestion
  /// @brief an output pin which will implement the behavior of an I2C scl pin
```

I prefer `///` for fields especially when the documentation is small. Also, this doesn't exactly mimic the SCL line, it is the SCL line, so lets call it as such.

Reply:
Same goes for the others.